src/haloniumpkg/utils.nim

```diff
@@ -113,20 +113,19 @@ proc `%`*(obj: tuple): JsonNode =
     result[field] = %val
 
 proc replace*(str: string, node: JsonNode): string =
-  let matches = str.findAll(re"\$(\w+)")
-  let keys = matches.mapIt(str[it.group(0)[0]])
-
-  var vals: seq[(string, string)]
-
-  for key in keys:
-    if not node.hasKey(key):
-      raise newException(URLTemplateException, fmt"Key '{key}' not found in JsonData")
-    if node[key].kind != JString:
-      raise newException(URLTemplateException, fmt"Key '{key}' is not a string")
-
-    vals.add(("$" & key, node[key].getStr()))
-
-  result = str.multiReplace(vals)
+  var keyVals: seq[(string, string)]
+  var isKey = false
+  for x in tokenize(str, {'$', '/'}):
```

Comment:
What is the slash here for?

Reply:
Perhaps I can explain it better with code:

```nim
import strutils

var keys: seq[string]
var found = false
for x in tokenize("/session/$sessionId/moz/xbl/$elementId/anonymous_by_attribute", {'$', '/'}):
  if x.isSep and x.token.endsWith('$'):
    found = true
  elif found and not x.isSep:
    found = false
    keys.add x.token

echo keys
```

It splits the path using both '$' and '/'

Reply:
Ah, yep, got it!